File: sagres/src/main/java/com/forcetower/sagres/database/dao/AccessDao.java
/*
 * Copyright (c) 2018.
 *
 * This file is part of the UNES Open Source Project.
 *
 * UNES is licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.forcetower.sagres.database.dao;

import com.forcetower.sagres.database.model.SagresAccess;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;

@Dao
public interface AccessDao {
    @Query("SELECT * FROM SagresAccess LIMIT 1")
    LiveData<SagresAccess> getAccess();

    @Query("SELECT * FROM SagresAccess LIMIT 1")
    SagresAccess getAccessDirect();

    @Insert
    void insert(SagresAccess access);
}

File: app/src/main/java/com/forcetower/uefs/core/storage/database/dao/AccessDao.kt
/*
 * This file is part of the UNES Open Source Project.
 * UNES is licensed under the GNU GPLv3.
 *
 * Copyright (c) 2020. João Paulo Sena <joaopaulo761@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.forcetower.uefs.core.storage.database.dao

import androidx.lifecycle.LiveData
import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy.REPLACE
import androidx.room.Query
import androidx.room.Transaction
import com.forcetower.uefs.core.model.unes.Access

@Dao
abstract class AccessDao {
    @Insert(onConflict = REPLACE)
    abstract fun insert(access: Access)

    @Transaction
    open fun insert(username: String, password: String) {
        val access = findDirect(username, password)
        if (access != null) return
        deleteAll()
        insert(Access(username = username, password = password))
    }

    @Query("SELECT * FROM Access WHERE username = :username AND password = :password LIMIT 1")
    abstract fun findDirect(username: String, password: String): Access?

    @Query("DELETE FROM Access")
    abstract fun deleteAll()

    @Query("SELECT * FROM Access LIMIT 1")
    abstract fun getAccess(): LiveData<Access?>

    @Query("SELECT * FROM Access LIMIT 1")
    abstract fun getAccessDirect(): Access?

    @Query("SELECT * FROM Access LIMIT 1")
    abstract suspend fun getAccessDirectSuspend(): Access?

    @Query("UPDATE Access SET valid = :valid")
    abstract fun setAccessValidation(valid: Boolean)

    @Query("UPDATE Access SET password = :password")
    abstract fun updateAccessPassword(password: String)

    @Query("UPDATE Access SET valid = :valid")
    abstract suspend fun setAccessValidationSuspend(valid: Boolean)

    @Query("UPDATE Access SET password = :password")
    abstract suspend fun updateAccessPasswordSuspend(password: String)

    @Query("SELECT COUNT(uid) FROM Access")
    abstract suspend fun getAccessCount(): Int
}
